app/ante/msg_gatekeeper.go

```diff
@@ -32,15 +33,35 @@ func (mgk MsgVersioningGateKeeper) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	if !exists {
 		return ctx, sdkerrors.ErrNotSupported.Wrapf("app version %d is not supported", ctx.BlockHeader().Version.App)
 	}
-	for _, msg := range tx.GetMsgs() {
+
+	if err := mgk.hasInvalidMsg(ctx, acceptedMsgs, tx.GetMsgs()); err != nil {
+		return ctx, err
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+func (mgk MsgVersioningGateKeeper) hasInvalidMsg(ctx sdk.Context, acceptedMsgs map[string]struct{}, msgs []sdk.Msg) error {
+	for _, msg := range msgs {
+		// Recursively check for invalid messages in nested authz messages.
+		if execMsg, ok := msg.(*authz.MsgExec); ok {
+			nestedMsgs, err := execMsg.GetMessages()
+			if err != nil {
+				return err
+			}
+			if err = mgk.hasInvalidMsg(ctx, acceptedMsgs, nestedMsgs); err != nil {
```

Comment:
do we want to put a limit on recursion? I'm not sure if this would be super valuable, but seems like good practice. I'm unaware of a usecase for authz that requires a bunch of nested msgs, but could be missing something.

if this wasn't consensus breaking before, it would be now

Reply:
What you could do is remove authz.MsgExec from the accepted messages list here. This would enforce that it's not possible to MsgExec a MsgExec which I don't think is useful and could have unknown repercussions if we allow it

Reply:
An alternative option to the one Callum proposed is:

```diff
func (mgk MsgVersioningGateKeeper) hasInvalidMsg(ctx sdk.Context, acceptedMsgs map[string]struct{}, msgs []sdk.Msg) error {
	for _, msg := range msgs {
		// Recursively check for invalid messages in nested authz messages.
		if execMsg, ok := msg.(*authz.MsgExec); ok {
			nestedMsgs, err := execMsg.GetMessages()
			if err != nil {
				return err
			}
+			if err = mgk.hasNestedMsgExec(ctx, nestedMsgs); err != nil {
+				return err
+			}
			if err = mgk.hasInvalidMsg(ctx, acceptedMsgs, nestedMsgs); err != nil {
				return err
			}
		}

		msgTypeURL := sdk.MsgTypeURL(msg)
		_, exists := acceptedMsgs[msgTypeURL]
		if !exists {
			return sdkerrors.ErrNotSupported.Wrapf("message type %s is not supported in version %d", msgTypeURL, ctx.BlockHeader().Version.App)
		}
	}

	return nil
}

+func (mgk MsgVersioningGateKeeper) hasNestedMsgExec(_ sdk.Context, nestedMsgs []sdk.Msg) error {
+	for _, nestedMsg := range nestedMsgs {
+		if _, ok := nestedMsg.(*authz.MsgExec); ok {
+			return sdkerrors.ErrNotSupported.Wrapf("nested MsgExec is not allowed in MsgExec")
+		}
+	}

	return nil
}
```

but my concern with adding this is that even though a MsgExec inside a MsgExec seems like a strange use case, what if it's actually desirable for certain scenarios. 

I don't see a major risk from not adding this constraint because if a user crafts a deeply nested MsgExec wrapper (i.e. `MsgExec(MsgExec(MsgExect(...)))` then the transaction will be rejected based on tx size.

Reply:
sgmt, that's likely better than a simple limit.

Reply:
I spun out an issue for this this discussion: https://github.com/celestiaorg/celestia-app/issues/3586